### PR TITLE
Photo to src

### PR DESCRIPTION
**Pull request by zeroadam**
Move Photo class to Friendica\Object namespace.
Imagick gave me a fit, but I'm no longer getting errors.

Related to #3878

### CODE REVIEW

File: src/Object/Photo.php [hunk 1]
```diff
@@ -0,0 +1,1206 @@
...
+	}
+
+	/**
+	 * @param object $data data
+	 * @return boolean
+	 */
+	private function loadData($data)
+	{
+		if ($this->isImagick()) {
+			$this->image = new \Imagick();
```
Comment (MrPetovan) [inline src/Object/Photo.php]: Why not add a couple of `use` statements? Wouldn't it remove the need for the horrible syntax like `$this->image->\Imagick::readImageBlob($data);`?

File: src/Object/Photo.php [hunk 2]
```diff
@@ -0,0 +1,1206 @@
...
+	 * @brief Updates the avatar links in a contact only if needed
+	 *
+	 * @param string $avatar Link to avatar picture
+	 * @param int    $uid    User id of contact owner
+	 * @param int    $cid    Contact id
+	 * @param bool   $force  force picture update
+	 *
+	 * @return array Returns array of the different avatar sizes
+	 */
+	public function updateContactAvatar($avatar, $uid, $cid, $force = false)
```
Comment (MrPetovan) [inline src/Object/Photo.php]: Can you move this to `\Friendica\Object\Contact::updateAvatar()` please?

File: src/Object/Photo.php [hunk 3]
```diff
@@ -0,0 +1,1206 @@
...
+			return false;
+		}
+	
+		if ($photo_failure) {
+			$photo = System::baseUrl() . '/images/person-175.jpg';
+			$thumb = System::baseUrl() . '/images/person-80.jpg';
+			$micro = System::baseUrl() . '/images/person-48.jpg';
+		}
+	
+		return(array($photo, $thumb, $micro));
```
Comment (MrPetovan) [inline src/Object/Photo.php]: Standards: Please remove parentheses from language constructs.

File: src/Object/Photo.php [hunk 4]
```diff
@@ -0,0 +1,1206 @@
...
+	}
+
+	/**
+	 * @param string $url url
+	 * @return object
+	 */
+	public function getPhotoInfo($url)
```
Comment (MrPetovan) [inline src/Object/Photo.php]: Please rename to `getInfoFromURL()`.

File: src/Object/Photo.php [hunk 5]
```diff
@@ -0,0 +1,1206 @@
...
+					$dest_width = $width;
+					$dest_height = $height;
+				}
+			}
+		}
+		return array("width" => $dest_width, "height" => $dest_height);
+	}
+
+	/**
+	 * @brief This function doesn't seem to be used
```
Comment (MrPetovan) [inline src/Object/Photo.php]: If the function isn't used anywhere, please scrap it. We will be able to find it again in the git history if needed. There already is a `Photo::store()` function anyway.

Comment (annando) [reply src/Object/Photo.php]: It is used in the addon "fromgplus".

Comment (MrPetovan) [reply src/Object/Photo.php]: What is the difference with `Photo::store()` ?

Comment (annando) [reply src/Object/Photo.php]: Good question. I guess we have to check the differences by comparing the functionality.

File: src/Object/Photo.php (hunk 1 above)
Comment (zeroadam) [reply src/Object/Photo.php]: Can you give me an example of how? I am having trouble getting this to work...

Comment (MrPetovan) [reply src/Object/Photo.php]: Just add 
```php
use Imagick;
use ImagickPixel;
```
After the other `use` statements, and you don't have to add the backslashes in front of the `Imagick` class names. Not sure about the individual method calls.

**Comment by zeroadam** [comment]
Phew! I think that did it... I don't know why I was having so much trouble with Imagick I had tried use statement early on but was getting errors, I think now it must have been related to something else. Should be all cleaned up now, and I'm back to no errors!

File: src/Object/Photo.php [hunk 6]
```diff
@@ -653,417 +758,244 @@ public function store($uid, $cid, $rid, $filename, $album, $scale, $profile = 0,
...
-			} else {
-				$dest_width = $width;
-				$dest_height = $height;
-			}
-		}
-	}
-	return array("width" => $dest_width, "height" => $dest_height);
-}
-
-function store_photo(App $a, $uid, $imagedata = "", $url = "") {
```
Comment (annando) [inline src/Object/Photo.php]: You removed this function, but you haven't changed the corresponding calls to this function in the file ```fromgplus.php```. Please make sure that either the call is changed or this function is restored. And no: I don't know the difference between the two functions that store photos.

**Comment by zeroadam** [comment]
Replaced the function and updated calls from addon. It looks like it was doing more than just storing the photo like in the store() function. I also fixed an error I got about using $this instead of self:: so If someone could look again and see if there are more of these that should be replaced, just in case!

File: src/Object/Photo.php [hunk 7]
```diff
@@ -653,417 +758,408 @@ public function store($uid, $cid, $rid, $filename, $album, $scale, $profile = 0,
...
+				 * Well, this not much better,
+				 * but at least it comes from the data inside the image,
+				 * we won't be tricked by a manipulated extension
+				 */
+				$image = new Imagick($filename);
+				$type = $image->getImageMimeType();
+				$image->setInterlaceScheme(Imagick::INTERLACE_PLANE);
+			} else {
+				$ext = pathinfo($filename, PATHINFO_EXTENSION);
+				$types = $this->supportedTypes();
```
Comment (MrPetovan) [inline src/Object/Photo.php]: This should be static as well.

File: src/Object/Photo.php [hunk 8]
```diff
@@ -653,417 +758,408 @@ public function store($uid, $cid, $rid, $filename, $album, $scale, $profile = 0,
...
-				}
+	/**
+	 * Guess image mimetype from filename or from Content-Type header
+	 *
+	 * @param string  $filename Image filename
+	 * @param boolean $fromcurl Check Content-Type header from curl request
+	 *
+	 * @return object
+	 */
+	public function guessImageType($filename, $fromcurl = false)
```
Comment (MrPetovan) [inline src/Object/Photo.php]: This function should be static.

File: src/Object/Photo.php [hunk 9]
```diff
@@ -653,417 +758,408 @@ public function store($uid, $cid, $rid, $filename, $album, $scale, $profile = 0,
...
-
-		$r = $img->store($uid, $cid, $hash, $filename, 'Contact Photos', 6);
+	/**
+	 * @param string  $photo         photo
+	 * @param integer $uid           user id
+	 * @param integer $cid           contact id
+	 * @param boolean $quit_on_error optional, default false
+	 * @return array
+	 */
+	public function importProfilePhoto($photo, $uid, $cid, $quit_on_error = false)
```
Comment (MrPetovan) [inline src/Object/Photo.php]: This function should be static.

File: src/Object/Photo.php [hunk 10]
```diff
@@ -653,417 +758,408 @@ public function store($uid, $cid, $rid, $filename, $album, $scale, $profile = 0,
...
-
-	$dest_width = $dest_height = 0;
-
-	if ((!$width) || (!$height)) {
-		return false;
+	/**
+	 * @param string $url url
+	 * @return object
+	 */
+	public function getInfoFromURL($url)
```
Comment (MrPetovan) [inline src/Object/Photo.php]: This function should be static.

File: src/Object/Photo.php [hunk 11]
```diff
@@ -653,417 +758,408 @@ public function store($uid, $cid, $rid, $filename, $album, $scale, $profile = 0,
...
-		} else {
-			$dest_width = intval(($width * $max) / $height);
-			$dest_height = $max;
+	/**
+	 * @param integer $width  width
+	 * @param integer $height height
+	 * @param integer $max    max
+	 * @return array
+	 */
+	public function scaleImageTo($width, $height, $max)
```
Comment (MrPetovan) [inline src/Object/Photo.php]: This function should be static.

File: src/Object/Photo.php [hunk 12]
```diff
@@ -653,417 +758,408 @@ public function store($uid, $cid, $rid, $filename, $album, $scale, $profile = 0,
...
 
+	/**
+	 * @brief This function is used by the fromgplus addon
+	 * @param object  $a         App
+	 * @param integer $uid       user id
+	 * @param string  $imagedata optional, default empty
+	 * @param string  $url       optional, default empty
+	 * @return array
+	 */
+	private function storePhoto(App $a, $uid, $imagedata = "", $url = "")
```
Comment (MrPetovan) [inline src/Object/Photo.php]: This function should be public and static.

**Comment by annando** [comment]
I'm now getting errors like this:
```
PHP Fatal error:  Uncaught ImagickException: no decode delegate for this image format `' @ error/blob.c/BlobToImage/353 in /var/www/virtual/pirati.ca/htdocs/src/Object/Photo.php:134
Stack trace:
#0 /.../src/Object/Photo.php(134): Imagick->readimageblob('<!DOCTYPE html\n...')
#1 /.../src/Object/Photo.php(76): Friendica\Object\Photo->loadData('<!DOCTYPE html\n...')
#2 /.../src/Object/Photo.php(843): Friendica\Object\Photo->__construct('<!DOCTYPE html\n...', 'image/jpeg')
#3 /.../src/Object/Contact.php(845): Friendica\Object\Photo::importProfilePhoto('https://gnusoci...', '1', '11355', true)
#4 /.../src/Protocol/OStatus.php(206): Friendica\Object\Contact::updateAvatar('https://gnusoci...', '1', '11355')
#5 /.../src/Protocol/OStatus.php(388): Friendica\Protocol\OStatus::fetchAuthor(Object(DOMXPath), NULL, Array, Array, false)
#6 /.../src/Object/Photo.php on line 134
PHP Fatal error:  Uncaught ImagickException: no decode delegate for this image format `' @ error/blob.c/BlobToImage/353 in /var/www/virtual/pirati.ca/htdocs/src/Object/Photo.php:134
Stack trace:
#0 /.../src/Object/Photo.php(134): Imagick->readimageblob('<!DOCTYPE html\n...')
#1 /.../src/Object/Photo.php(76): Friendica\Object\Photo->loadData('<!DOCTYPE html\n...')
#2 /.../src/Object/Photo.php(843): Friendica\Object\Photo->__construct('<!DOCTYPE html\n...', 'image/jpeg')
#3 /.../src/Object/Contact.php(845): Friendica\Object\Photo::importProfilePhoto('https://gnusoci...', 0, '9914', true)
#4 /.../src/Object/Contact.php(649): Friendica\Object\Contact::updateAvatar('https://gnusoci...', 0, '9914')
#5 /.../src/Protocol/OStatus.php(210): Friendica\Object\Contact::getIdForURL('https://gnusoci...', 0)
#6 /.../src/Protocol/OStatus.php(38 in /var/www/virtual/pirati.ca/htdocs/src/Object/Photo.php on line 134
PHP Fatal error:  Uncaught ImagickException: no decode delegate for this image format `' @ error/blob.c/BlobToImage/353 in /var/www/virtual/pirati.ca/htdocs/src/Object/Photo.php:134
Stack trace:
#0 /.../src/Object/Photo.php(134): Imagick->readimageblob('<!DOCTYPE html\n...')
#1 /.../src/Object/Photo.php(76): Friendica\Object\Photo->loadData('<!DOCTYPE html\n...')
#2 /.../src/Object/Photo.php(843): Friendica\Object\Photo->__construct('<!DOCTYPE html\n...', 'image/jpeg')
#3 /.../src/Object/Contact.php(845): Friendica\Object\Photo::importProfilePhoto('https://indy.im...', 0, '499513', true)
#4 /.../src/Object/Contact.php(649): Friendica\Object\Contact::updateAvatar('https://indy.im...', 0, '499513')
#5 /.../src/Protocol/OStatus.php(210): Friendica\Object\Contact::getIdForURL('https://gnusoci...', 0)
#6 /.../src/Protocol/OStatus.ph in /var/www/virtual/pirati.ca/htdocs/src/Object/Photo.php on line 134
```

**Comment by zeroadam** [comment]
This looks like html data being sent in?

**Comment by MrPetovan** [comment]
Yep, the gnusocial URL returns plain HTML when it expects an image.

**Comment by annando** [comment]
The problem is not that the data is bad - but that this here doesn't work: https://github.com/friendica/friendica/blob/develop/src/Object/Photo.php#L133-L138

```try``` ... ```catch``` should catch this exception, shouldn't it?

**Comment by MrPetovan** [comment]
It’s a namespace issue then, `\ImagickException` extends `\Exception`, not
`Friendica\Object\Exception`, we should either add `use Exception;` or use
`\Exception`

**Comment by zeroadam** [comment]
https://github.com/friendica/friendica/pull/3992

**Comment by MrPetovan** [comment]
👌👍

**Comment by annando** [comment]
I guess that we will have this issue at other places in the code as well.

**Pull request merged**